### ISSUE TITLE
Attempting to train Cyrillic based language: Mongolian and findings

### ISSUE THREAD

**Issue by amarbayar** [SWivid/F5-TTS#464]
### Checks

- [X] This template is only for question, not feature requests or bug reports.
- [X] I have thoroughly reviewed the project documentation and read the related paper(s).
- [X] I have searched for existing issues, including closed ones, no similar questions.
- [X] I confirm that I am using English to submit this report in order to facilitate communication.

### Question details

After reading some successful cases and promising results (like the French model release in https://github.com/SWivid/F5-TTS/issues/434 by @RASPIAUDIO, Youtube videos posted by @JarodMica) I have decided to give it a  try to train a model for the Mongolian language. Mongolian language is based off of Cyrillic, so has a similar alphabet to Russian but has 2 additional letters. Pronunciation differs, though. And coming from software engineering background, without much experience on building ML models, has been posing the fair share of challenges. I am creating this issue to provide what I have tried and ask questions for guidances from folks who might be able to guide me in the right direction as I am not seeing satisfactory results so far. 

### Hardware specs:
- Renting an hourly cloud GPU: 1 x A40 9 vCPU 50 GB RAM

### Dataset: 
- 3800~ samples of a single speaker in studio quality recording of about 6 hours
- Sample rate at 24,000 Hz. Examples:
```
af1|Гэрэл сайн болсныг Бурхан харж, Бурхан гэрлийг харанхуйгаас салгажээ.
af2|Бурхан гэрлийг өдөр гэж нэрлээд, харанхуйг шөнө гэж нэрлэлээ.
```

### Gradio App
Followed the Installation steps, downloaded and installed the dependencies/packages/libraries and ran the Gradio app for fine tuning. 

- Loaded my dataset
- Did a vocab check and it found the 2 additional letters as expected: `ө, ү` so extended the vocab with them
- Tokenizer example:
```
['   ' , ' н ' , '   ' , ' э ' , '   ' , ' г ' , '   ' , ' и ' , '   ' , ' й ' , '   ' , ' н ' , '   ' , ' х ' , '   ' , '   ' , ' н ' , '   ' , ' ь ' , '   ' , '   ' , ' н ' , '   ' , ' э ' , '   ' , ' р ' , '   ' , '   ' , ' а ' , '   ' , ' д ' , '   ' , ' а ' , ' , ' , '   ' , '   ' , ' н ' , '   ' , ' ө ' , '   ' , ' г ' , '   ' , ' ө ' , '   ' , ' ө ' , '   ' , ' г ' , '   ' , ' и ' , '   ' , ' й ' , '   ' , ' н ' , '   ' , ' х ' , '   ' , '   ' , ' н ' , '   ' , ' ь ' , '   ' , '   ' , ' н ' , '   ' , ' э ' , '   ' , ' р ' , '   ' , '   ' , ' з ' , '   ' , ' и ' , '   ' , ' л ' , '   ' , ' л ' , '   ' , ' а ' , ' . ']
```

### The following are various scenarios I have tried. 

- Tokenizer type: `pinyin` and also tried `char`
- Batch size per GPU: 3200, 6400, 9200, 19200
- Max samples: 32, 64
- Learning Rate: tried 0.00001 and also 0.000005
- Model: F5TTS_Base
- Gradient Accumulation Steps: 1 (didn't change it)
- Max Gradient Norm: 1 (didn't change it)
- Epoch: tried 50, 100, 150
- Warmup Updates: tried 100, 300, 500, 1000
- Mixed Precision: fp16 (as I think A40 supports it)
- Note: I did try the `Auto Settings` as well. I think the batch size per GPU was `1745` which is too small for my GPU? And epoch came up to be `781`. 

### Results so far
I have been testing the checkpoints at various intervals and haven't gotten satisfactory results. The voice does sound very similar to the original sound of the speaker provided in the dataset. But:

1. Words are jibberish
2. It is too fast (with no breaks...etc) as if it is trying a tongue-twister. 

### Loss 
I think the results I am seeing aligns well with the Loss graph that I see on Tensorboard as well. Below is from my latest training attempt with 19,200 batch size per GPU, 32 max samples, 50 epochs, 300 warmup updates, 400 save per updates, 200 last per steps, fp16, f5tts_base model, `char` tokenizer, learning rate of `0.00001` and batch size type: `frame`. 

Loss Rate: 
<img width="677" alt="Screenshot 2024-11-14 at 10 31 40 AM" src="https://github.com/user-attachments/assets/941c1d2d-ad3b-414a-9798-6836f25a4945">

Learning Rate Decay:
<img width="644" alt="image" src="https://github.com/user-attachments/assets/cfd0d596-f6e7-4828-a3fd-42f97aa03817">

Reference audio: https://whyp.it/tracks/225389/af143?token=s8DdA
Checkpoint inference: https://whyp.it/tracks/225391/af143-ckpt?token=i4bi0

### Thoughts / Questions 
I have read in a few issues where @SWivid  for example mentions that at 200K steps or beyond, we might get some intelliglbe audio. But I have also seen examples where @RASPIAUDIO  in his Youtube video tutorial had roughly 200-300 samples and the sample checkpoint was not too bad and there was an improvement which could then be a trigger to collect more dataset and spend more time training. In my case, I am not seeing much improvements even with nearly 4000 samples (I know it is small) and 6 hours length of a single speaker in studio quality. At least if I hear some words proncounced correctly or close enough, proper gaps and pauses then I could look at more datasets...etc. 

So, could it be that:
1. There is an inherent issue with the language itself?
2. Or possible issue with cyrillic? (I do see the vocab.txt already had Cyrillic symbols) 
3. Should I just not expect to see loss decrease at 7000th step mark and just aim to get 200K steps at min?
4. Should I try different settings? (I did try the `Auto Settings` as well) 
5. Or should I keep adding more and more data set and iterate this? 

Any of these could be like a shot in the dark and I am trying to understand if there is a systematic way to approach it so that I can rule-in or rule-out hypothesis early on?

**Comment by SWivid** [comment]
> 200K steps or beyond

it's for training from scratch, say pretraining; finetuning will allow fewer steps to hear something.
As Cyrillic is somehow different from pretrained distribution, e.g. we know French and English is close to each other in some way, it would take more time to learn pronunciation.

Possible improvement:
1. grapheme-to-phoneme tokenization
2. change the tokenizer like `"Г", "э", "р", "э", "л", " ", "с", ...,]`, having words together

**Comment by JarodMica** [comment]
> * Tokenizer example:
> 
> ```
> ['   ' , ' н ' , '   ' , ' э ' , '   ' , ' г ' , '   ' , ' и ' , '   ' , ' й ' , '   ' , ' н ' , '   ' , ' х ' , '   ' , '   ' , ' н ' , '   ' , ' ь ' , '   ' , '   ' , ' н ' , '   ' , ' э ' , '   ' , ' р ' , '   ' , '   ' , ' а ' , '   ' , ' д ' , '   ' , ' а ' , ' , ' , '   ' , '   ' , ' н ' , '   ' , ' ө ' , '   ' , ' г ' , '   ' , ' ө ' , '   ' , ' ө ' , '   ' , ' г ' , '   ' , ' и ' , '   ' , ' й ' , '   ' , ' н ' , '   ' , ' х ' , '   ' , '   ' , ' н ' , '   ' , ' ь ' , '   ' , '   ' , ' н ' , '   ' , ' э ' , '   ' , ' р ' , '   ' , '   ' , ' з ' , '   ' , ' и ' , '   ' , ' л ' , '   ' , ' л ' , '   ' , ' а ' , ' . ']
> ```

This is probably 99% the issue right here.  You'll want to modify `convert_to_pinyin` to not add spaces between Cyrillic.  I saw the same issue with jumbled words in Japanese because it wants to add the extra space:

Sample modification:
```
def convert_char_to_pinyin(text_list, polyphone=True):
    final_text_list = []
    zh_quote_trans = str.maketrans({'“': '"', '”': '"', '‘': "'", '’': "'"})
    custom_trans = str.maketrans({';': ','})
    
    def is_japanese(c):
        return (
            '\u3040' <= c <= '\u309F' or  # Hiragana
            '\u30A0' <= c <= '\u30FF' or  # Katakana
            '\uFF66' <= c <= '\uFF9F'     # Half-width Katakana
        )
    
    for text in text_list:
        char_list = []
        text = text.translate(zh_quote_trans)
        text = text.translate(custom_trans)
        for seg in jieba.cut(text):
            seg_byte_len = len(seg.encode('utf-8'))
            if seg_byte_len == len(seg):
                if char_list and seg_byte_len > 1 and char_list[-1] not in " :'\"":
                    char_list.append(" ")
                char_list.extend(seg)
            elif polyphone and seg_byte_len == 3 * len(seg):
                seg_pinyin = lazy_pinyin(seg, style=Style.TONE3, tone_sandhi=True)
                for p in seg_pinyin:
                    if p not in "。，、；：？！《》【】—…":
                        if not char_list or not is_japanese(char_list[-1]):
                            char_list.append(" ")
                    char_list.append(p)
            else:
                for c in seg:
                    if ord(c) < 256:
                        char_list.append(c)
                    elif is_japanese(c):
                        char_list.append(c)
                    else:
                        if c not in "。，、；：？！《》【】—…":
                            if not char_list or not is_japanese(char_list[-1]):
                                char_list.append(" ")
                            pinyin = lazy_pinyin(c, style=Style.TONE3, tone_sandhi=True)
                            char_list.extend(pinyin)
                        else:
                            char_list.append(c)
        final_text_list.append(char_list)
    return final_text_list
```

**Comment by amarbayar** [comment]
Okay. Thanks to both of you. Will try in the following order and see what happens. If I run into any issues or make good progress, I will share my findings here. 

1. Remove the logic that is adding additional space between letters/symbols if it is not Chinese such that tokenized text is represented like:
```
['   ' , ' ё ' , '   ' , ' с ' , '   ' , ' о ' , '   ' , ' л ' , '   ' , ' с ' , '   ' , ' о ' , '   ' , ' н ' , '   ' , ' т ' , '   ' , ' о ' , '   ' , ' й ' , '   ' , ' г ' , '   ' , ' о ' , '   ' , ' о ' , '   ' , '   ' , ' д ' , '   ' , ' о ' , '   ' , ' л ' , '   ' , ' о ' , '   ' , ' о ' , '   ' , '   ' , ' х ' , '   ' , ' о ' , '   ' , ' н ' , '   ' , ' о ' , '   ' , ' г ' , '   ' , ' и ' , '   ' , ' й ' , '   ' , ' г ' , '   ' , '   ' , ' г ' , '   ' , ' ү ' , '   ' , ' й ' , '   ' , ' ц ' , '   ' , ' э ' , '   ' , ' э ' , ' . ']
```

2. If this doesn't yield good results, will try the phoneme approach. I. have found: https://github.com/xinjli/transphone. The rough idea I have is that I can use it to phonemize my dataset text, extract the unique symbols and add them to the pretrained vocab by extending. And I will need to implement a logic that converts each cyrillic character to the phoneme so that it can do a lookup from the vobab.txt. But will get to it after I experiment the first / faster approach.

**Comment by amarbayar** [comment]
Alright some progress. 

After extending the pre-trained vocab.txt with the additional letters from MN language, I made some tweak to the `convert_char_to_pinyin()` method so that it doesn't add those spaces between every two characters and even multiples of those spaces at times.

Exampled tokenized text:
```
[' а ' , ' х ' , '   ' , ' д ' , ' ү ' , ' ү ' , '   ' , ' н ' , ' а ' , ' р ' , ' ы ' , ' н ' , ' х ' , ' а ' , ' а ' , '   ' , ' э ' , ' з ' , ' э ' , ' н ' , '   ' , ' н ' , ' ь ' , '   ' , ' ч ' , ' и ' , '   ' , ' б ' , ' о ' , ' л ' , ' ж ' , ' , ' , '   ' , ' э ' , ' х ' , ' и ' , ' й ' , ' н ' , '   ' , ' ч ' , ' и ' , ' н ' , ' ь ' , '   ' , ' х ' , ' ө ' , ' в ' , ' г ' , ' ү ' , ' ү ' , ' д ' , '   ' , ' ч ' , ' а ' , ' м ' , ' д ' , '   ' , ' с ' , ' ө ' , ' г ' , ' д ' , ' ө ' , ' н ' , '   ' , ' м ' , ' ө ' , ' р ' , ' г ' , ' ө ' , ' х ' , '   ' , ' б ' , ' о ' , ' л ' , ' т ' , ' у ' , ' г ' , ' а ' , ' й ' , ' . ']
```

Looked much better. 

So then, I tried with:
```
Batch Size Type: Frame
Finetune: checked
Model: F5TTS_Base
Learning Rate: 0.00001
Batch Size per GPU: 19200
Max Samples: 64
Gradient Accumulation Steps: 1
Max Gradient Norm: 1
Epoch: 100
Warmup Updates: 300
Save per updates: 1200
Last per Steps: 1000
Mixed Precision: none
``` 

Even at around 15th epoch, it was still not doing well. The sampled audio was jibberish and inference also output jibberish. And the Loss Rate had plateaued at around 5th epoch with no further significant improvements. 

Tried a few different settings on batch size per GPU, warmup updates. And still didn't have much luck. 

Then I decided to give the `Auto Settings` a try. It recommended: 
```
Batch Size Type: Frame
Finetune: checked
Model: F5TTS_Base
Learning Rate: 0.00001
Batch Size per GPU: 1745
Max Samples: 64
Gradient Accumulation Steps: 1
Max Gradient Norm: 1
Epoch: 781
Warmup Updates: 192
Mixed Precision: none
```

So I just took it. The `Save Per Updates` and `Last per Steps` were too low and I didn't want fill my desk with tons of 5GB+ checkpoints (and yep, I am ware of the model reduction capability which brings it down to 1.3GB which is great but didn't want to do it everytime or code to automate it just yet). So I bumped these settings up to 1200 and 1000 respectively. 

Overall, it has been doing better than all of my previous attempts:

<img width="636" alt="image" src="https://github.com/user-attachments/assets/120971d2-15d4-4627-9f81-1f1bdc73fe7e">

And the nice thing is that the sample generated audio files at checkpoints has been getting better. 

### However...

I am not having a hard time getting the inference to produce similar output as the sample generated audio and it is not happening... In other words, a sample audio generated at `step_30000.pt` sounds pretty good. So then when I run something like:
```
f5-tts_infer-cli \
--model "F5-TTS" \
--ckpt_file ckpts/mn_tts/model_30000.pt \
--vocab_file data/mn_tts_char/vocab.txt \
--ref_audio "data/mn_tts_char/wavs/af1.wav" \
--ref_text "Энэ бол тест өгүүлбэр." \
--gen_text "Энэ бол тест өгүүлбэр."
```

and the toml:
```
model = "F5-TTS"
ref_audio = "data/mn_tts_char/wavs/af1.wav"  
ref_text = "Энэ бол тест өгүүлбэр."
gen_text = "Энэ бол тест өгүүлбэр."
gen_file = ""
remove_silence = false
output_dir = "tests"
nfe_step = 64
cfg_strength = 2.0
sway_sampling = -1
use_ema = true
```

I tried a few different settings here. The output sounds very very bad to a point where I cannot make any sense out of it. However, the sample audio generated at step_30000 in the Gradio App sounds good. 

<img width="953" alt="image" src="https://github.com/user-attachments/assets/e2bdd3d5-af8f-43ca-88ea-bb6ebaec93ec">

@SWivid  -- Do you know why a inference generated audio might sound way different (much worse) than sample generated audio at checkpoints? 

@JarodMica -- Wondering if you ran into something similar?

**Comment by JarodMica** [comment]
It's hard to compare steps between different setups, but I think your model still needs much more training. I know some of my models at earlier points in training were messing up speech but I didn't test it too throughly, so I don't know what the "minimum" training you'd need to do for a language to be coherent.

For reference, the model I showed in my video was trained for at least 180k steps at a batch size of 4000, so I would say it saw the target audio about 12x more than your output at 30k steps (My mental math comes from 1700x30000 vs 4000x180000).

**Comment by amarbayar** [comment]
Okay. I have been playing with the inference parameters and have gotten to a point where the inferenced audios are somewhat close to the sample generated audios that got created during training. Still there is some difference but better than before. 

In case it is helpful for others... I have written down my findings based on my understanding (so caution is that they might not be 100% accurate, but if there is a framework like this to use to tweak the settings it seems a bit easier than just trying a bunch of different numbers in the dark)

### Audio Generation Parameters
| Parameter                                  | Description                                                                                                                                      | Code Example            | Default Value | Notes                                                                                                                                                                                     |
|--------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------|---------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NFE (Number of Function Evaluations) Steps | Controls the number of steps the model takes to generate audio from noise. Higher values potentially yield better quality but slower generation. | nfe_step = 32           | 32            | Defined in **utils_infer.py**. Affects the ODE solver's granularity in **cfm.py**: `trajectory = odeint(fn, y0, t, **self.odeint_kwargs)`                                                           |
| CFG (Classifier-Free Guidance) Strength    | Controls how strongly the model adheres to the conditioning (reference audio and text).                                                          | cfg_strength = 2.0      | 2.0           | Found in **cfm.py**. Higher values enforce stronger adherence to the reference style but may sound artificial. Lower values sound more natural but might deviate from the reference.          |
| Sway Sampling Coefficient                  | Affects the trajectory of noise-to-audio generation.                                                                                             | sway_sampling_coef = -1 | -1            | Implemented in **cfm.py**: `if sway_sampling_coef is not None: t = t + sway_sampling_coef * (torch.cos(torch.pi / 2 * t) - 1 + t)`. Negative values indicate more aggressive sampling early on. |
| EMA (Exponential Moving Average)           | Controls whether to use averaged model weights or direct weights.                                                                                | use_ema = True          | True          | Implemented in `utils_infer.py`. EMA weights offer more stability but might not reflect recent training enhancements.                                                                       |

### Output Control Parameters
| Parameter                     | Description                                                                           | Code Example               | Default Value | Notes                                                                                        |
|-------------------------------|---------------------------------------------------------------------------------------|----------------------------|---------------|----------------------------------------------------------------------------------------------|
| Target RMS (Root Mean Square) | Controls audio volume normalization.                                                  | target_rms = 0.1           | 0.1           | Implemented in **utils_infer.py**. Higher values result in louder output.                        |
| Cross-fade Duration           | Used when joining multiple audio segments, influencing the smoothness of transitions. | cross_fade_duration = 0.15 | 0.15 seconds  | Crucial for long text generation requiring chunking. Implemented in **utils_infer.py**.          |
| Speed                         | Controls the speaking rate of the generated audio.                                    | speed = 1.0                | 1.0           | Affects duration calculation in **utils_infer.py**. Values greater than 1 produce faster speech. |

**Comment by SWivid** [comment]
> adding data to the existing dataset is a good practice

It' 's better, would help avoid catastrophic forgetting

**Comment by JarodMica** [comment]
Based on my understanding, we're training all the weights of the model so by adding new data and training on new data, we eventually move away from any previously trained data towards our current training dataset so you'd want to keep your previous dataset in there. 

My current Japanese model I did the same thing and just continued training with the new dataset, though, I reinitialized LR from scratch and optimizer schedule

**Comment by amarbayar** [comment]
@JarodMica  - How are you reinitializing your LR? I think the settings are saved in the `ckpts/{project}/setting.json`. Are you clearing it or are you just overriding the settings in the Gradio app? 

And I haven't seen an `optimizer schedule`. What/where is it? (I did read about it briefly when I was doing some research regarding the learning rate but didn't understand it much)

**Comment by SWivid** [comment]
> reinitializing your LR

just reduce/prune the last.pt and rerun the training/finetune; that will work to reinitialize the lr as i glance over the script

**Comment by JarodMica** [comment]
> @JarodMica - How are you reinitializing your LR? I think the settings are saved in the `ckpts/{project}/setting.json`. Are you clearing it or are you just overriding the settings in the Gradio app?
> 

My training script is old and messy, the easiest way is what @SWivid recommends so that you load the model with EMA weights.

I temporarily commented out this 181-183 before starting the run, then uncommented: https://github.com/SWivid/F5-TTS/blob/8f65f9f3e4d47611f2c0628cb3732b1b1ba3fd0b/src/f5_tts/model/trainer.py#L180

**Comment by SWivid** [comment]
[off topic]

@JarodMica we actually got a [more open place](https://github.com/SWivid/F5-TTS#:~:text=Custom%20model%20inference%20(local%20only)) for model sharing: 
[src/f5_tts/infer/SHARED.md](https://github.com/SWivid/F5-TTS/blob/main/src/f5_tts/infer/SHARED.md), while the lastest version of local infer_gradio support loading custom ckpts.

Welcome to share if willing to when have time~ 
Thought you've got a nice JA TTS, just not sure if the shared.md and custom loading feature is not conveyed or refinement of model is still ongoing, lol~

**Comment by JarodMica** [comment]
> [off topic]
> 
> @JarodMica we actually got a [more open place](https://github.com/SWivid/F5-TTS#:~:text=Custom%20model%20inference%20(local%20only)) for model sharing: [src/f5_tts/infer/SHARED.md](https://github.com/SWivid/F5-TTS/blob/main/src/f5_tts/infer/SHARED.md), while the lastest version of local infer_gradio support loading custom ckpts.
> 
> Welcome to share if willing to when have time~ Thought you've got a nice JA TTS, just not sure if the shared.md and custom loading feature is not conveyed or refinement of model is still ongoing, lol~

Oh perfect!  I'll be sharing the JA model when it finishes up, but the intermediary results might be satisfactory for most right now.  I launched up training again because I reached the end of my LR and it didn't seem like it was making any progress.  

I'll have to create a new branch to push some updates and changes lol, just haven't had the time since my main branch is incompatible

**Comment by amarbayar** [comment]
Updated dataset and fine-tuning isn't doing very well... Also, tried bunch of different parameters. The new delta data is processed, cleaned up...etc but the new dataset contains multiple people voice while the previous bulk training was only one person. Not sure if this would cause an issue... 

<img width="642" alt="image" src="https://github.com/user-attachments/assets/dea5fb7e-62c6-4eaf-8eaa-51af8bf1868e">

**Comment by wanfnc** [comment]
@amarbayar 
Hi, I tried training Korean model recently, and reached here. I just wonder how is it going now?
Is your model work well?

**Comment by amarbayar** [comment]
Hi @wanfnc  -- I couldn't break through loss rate of 0.60~ regardless of hyper parameter changes (LR, Batch Size, Grad Accum Steps, Clip Size and even tried CosineAnnealing LR optimizer...etc). I even tried augmenting my dataset, and went to get additional 100~ hours of dataset and loss rate didn't really improve beyond 0.60~.

**Comment by wanfnc** [comment]
Thanks for reply @amarbayar !
Then how about your tts outputs?
I still training model but it takes so looooooonnng ( cause my GPU 3080 10GB )
I tested with model_last.pt ( 67000 steps of 8GB dataset ) about 3 and half epochs,
my output still speak like Chinese.
Have you got similar issue? I mean the infer outputs are Chinese.

**Comment by amarbayar** [comment]
@wanfnc  - I think 67,000 steps is somewhat on the lower end. While I didn't hear "Chinese" speeches, the inferenced output wasn't intelligible at 50K-100K. But towards 100K and beyond, I started hearing pretty good, intelligible output. But I had to keep playing with the parameters and kept building on top of the latest good checkpoint to continue finetuning the effort. 

If you are hearing "Chinese", it sounds like you are fine-tuning instead of training from scratch which I think is good. But I would focus more on the stats like Training Loss Rate, which should be an indicator to help you assess whether the model is learning and improving or not. Are you running tensorboard and are you monitoring the Loss Rate graph? Also, another crucial step is to make sure your vocab is created properly. It looks like the "Emilia" vocab already contains Korean vocabs:
- https://github.com/SWivid/F5-TTS/blob/main/data/Emilia_ZH_EN_pinyin/vocab.txt#L2043-L2058
- https://github.com/SWivid/F5-TTS/blob/main/data/Emilia_ZH_EN_pinyin/vocab.txt#L2060-L2538

So, I would think that the pre-trained model might already be exposed to general Korean language dataset and might have learned the basic patterns. And I would assume that fine-tuning on high quality dataset might yield pretty good results while in my case, MN is a pretty complex and "low-resource" language so it took a while to get the vocabs right, gather dataset...etc.

**Comment by wanfnc** [comment]
@amarbayar 
I find a suspected cause. Test model didn't work in my fine-tuning gradio WEBUI, that's way I tested my model in f5-tts at pinokio. 
And I found below code in ["preapare_emilia.py".
](https://github.com/SWivid/F5-TTS/blob/0297be2541f9a062f9d54103926bdb88d63440ea/scripts/prepare_emilia.py#L128C5-L128C25)


**if __name__ == "__main__":

    max_workers = 32

    tokenizer = "pinyin"  # "pinyin" | "char"
    polyphone = True

    langs = ["ZH", "EN"]
    dataset_dir = "<SOME_PATH>/Emilia_Dataset/raw"
    dataset_name = f"Emilia_{'_'.join(langs)}_{tokenizer}"
    print(f"\nPrepare for {dataset_name}\n")

    main()**


Is it means infering tts is only do work in Chinese and English? or I must do something or change code before training?

Actually, I follow training steps from Youtube by @JarodMica [Video](https://youtu.be/GmketyZW2c4?si=iUouwRNaAyxuo6Bg)
Is there any missig steps here?

Before finish a comment, I apologize for my shallow knowledge in computer engineering.

So much thanks guys!

**Comment by amarbayar** [comment]
@wanfnc - Do you have your own dataset? I think the Emilia dataset is only applicable for EN and ZH. If you have your own dataset, you should use it and prepare the:
- `metadata.csv` file (containing audio file names and their transcriptions separated by `|`)
- `wavs/` folder containing the .wav audio files sampled to the same Herz i.e 16,000
Once you have your dataset, you can simply open the Gradio web app, create a new project (I am assuming for Korean, you might need the `pinyin` tokenizer? I used `char` for Cyrillic). Then you will need to place your `wavs/` dir and `metadata.csv` file in your `data/{project_name}_{tokenizer}/` folder and try to Load the Data to see if the Gradio app can recognize your dataset. If it recognizes it then it will tell you if it generated the vocab or not and if there are any vocabs missing...etc. 

I guess without knowing what you have done exactly, what dataset you have, how you prepared your dataset, whether your created a project...etc it is kind of shooting in the dark to try to guess what might be happening.

**Comment by wanfnc** [comment]
@amarbayar 
Hi! I'm really appreciate for your kindness. My PC and GPU worked 2 days and nights, finally I got some intelligible results!
You're right, the problem was cause of lack of training steps and epochs.

You might be wondering,
I have my own dataset, and @JarodMica 's video has a step of 'Transcribe data' and 'Vocab Check', that makes me a metadata.csv and vocab.txt. I'm not sure that data is 100% correct but was no time for check and fix it.
My dataset is about 8GB, 44kHz, mono wav file. ( I guess about 24 hours ) I also used `char` for Korean.

Maybe I have to test several times to make model to achieve high quality.

**Comment by Alykasym** [comment]
Hi @amarbayar 
Thanks so much for sharing your findings, they were really helpful.

I wanted to ask about an issue I’ve been struggling with. During training, the samples generated while saving checkpoints sound great, but when I run actual inference, the results are noticeably worse.

I noticed you mentioned some pointers about the CFM, but I’d love to hear more specific insights if you could share.

**Comment by amarbayar** [comment]
@Alykasym  - Yeah, I think I did notice that behavior as well. Then I realized a few things:
1. The audio generation parameters had a big impact in terms of actual inference quality. https://github.com/SWivid/F5-TTS/issues/464#issuecomment-2477993364
2. The reference audio and text must be of similar data you used for your training (audio length-wise, normalized text-wise)
3. Remember to normalize your text when you synthesize speech. If your input text is different than your normalized text pattern used during training, it might not work. 

My understanding is that CFM helps with "mimicking" the reference voice.
